# Question 1 Algorithmic questions
Find the kth largest element in an unsorted array. Note that it is the kth largest element in the sorted order, not the kth distinct element.

In [1]:
import heapq
def kthlargest(array,k):
    """function returns the kth largest element in an array
       assume len(array) is n and k is always valid,
       overall time complexity is O(n+k*log(n))
    """
    heap = [-i for i in array]    #->time complexity: O(n)
    heapq.heapify(heap)           #->time complexity: O(n)
    for _ in range(k-1):          
        heapq.heappop(heap)       #->each heapq.heappop is O(log(n))
    return -1*heapq.heappop(heap)

In [5]:
array1,k1 = [3,2,1,5,6,4],2         #test case 1&2: unsorted array but contains only distinct values
array2,k2 = [11,73,13,46,91,35],4
array3,k3 = [3,2,3,1,2,4,5,5,6],5   #test case 3&4: unsorted array that contains duplicate values
array4,k4 = [44,23,76,23,65,23,44,89,12],6

print(kthlargest(array1,k1))
print(kthlargest(array2,k2))
print(kthlargest(array3,k3))
print(kthlargest(array4,k4))

5
35
3
23


# Question 2 Multithreading pop
Using python data structure Queue, generate a queue that have 10000 elments. Then use 1, 5, 20 threads to pop out the elments.

In [6]:
import threading, queue
q = queue.Queue()
for task in range(10000):
    q.put(task)

def thread_function(name,remaining_thread):
    tasks=q.qsize()
    print(f'Thread {name}: starting')
    for i in range(tasks//remaining_thread):
        q.get()
    print(f'Thread {name}: finished')


In [7]:
%%time
if __name__ == "__main__":
    print("Main: start running thread")
    print(f'Total tasks in the queue: {q.qsize()}')
    threads = list()
    indexs,remaining_thread = [1,5,20],3
    for each in indexs:
        x = threading.Thread(target=thread_function, args=(each,remaining_thread))
        threads.append(x)
        x.start()
        x.join()
        remaining_thread -= 1
    print(f'Remaing tasks in the queue: {q.qsize()}')
    print("Main: all done\n")

Main: start running thread
Total tasks in the queue: 10000
Thread 1: starting
Thread 1: finished
Thread 5: starting
Thread 5: finished
Thread 20: starting
Thread 20: finished
Remaing tasks in the queue: 0
Main: all done

CPU times: user 30.4 ms, sys: 2.92 ms, total: 33.4 ms
Wall time: 31.1 ms


### Observations:  
Using 3 threads- CPU times: user 30.4 ms, sys: 2.92 ms, total: 33.4 ms, Wall time: 31.1 ms  
Using 4 threads- CPU times: user 30.7 ms, sys: 2.74 ms, total: 33.4 ms, Wall time: 31.5 ms  
Using 5 threads- CPU times: user 31.2 ms, sys: 3.39 ms, total: 34.6 ms, Wall time: 31.9 ms  
Using 7 threads- CPU times: user 30.8 ms, sys: 3.27 ms, total: 34 ms, Wall time: 31.7 ms  
Using 12 threads- CPU times: user 31.1 ms, sys: 3.79 ms, total: 34.9 ms, Wall time: 32.5 ms  
### Explanation:  
It is easy to see that the CPU/total/Wall times for different numbers of thread are all very close.  
This shows that using more threads does not spead up finishing all tasks and it is because Python Global Interpreter Lock (GIL) limits one Python thread to run at a time.